scene/resources/skeleton_modification_3d_fabrik.h

```diff
@@ -0,0 +1,121 @@
+/* Copyright (c) 2014-2021 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "scene/3d/skeleton_3d.h"
+#include "scene/resources/skeleton_modification_3d.h"
+
+#ifndef SKELETONMODIFICATION3DFABRIK_H
+#define SKELETONMODIFICATION3DFABRIK_H
+
+class SkeletonModification3DFABRIK : public SkeletonModification3D {
+	GDCLASS(SkeletonModification3DFABRIK, SkeletonModification3D);
+
+private:
+	struct FABRIK_Joint_Data {
+		String bone_name = "";
+		int bone_idx = -1;
+		float length = -1;
+		Vector3 magnet_position = Vector3(0, 0, 0);
+
+		bool auto_calculate_length = true;
+		bool use_tip_node = false;
+		NodePath tip_node = NodePath();
+		ObjectID tip_node_cache;
+
+		bool use_target_basis = false;
+		float roll = 0;
+	};
+
+	Vector<FABRIK_Joint_Data> fabrik_data_chain;
+	NodePath target_node;
+	ObjectID target_node_cache;
+
+	float chain_tolerance = 0.01;
+	int chain_max_iterations = 10;
+	int chain_iterations = 0;
+
+	void update_target_cache();
+	void update_joint_tip_cache(int p_joint_idx);
+
+	int final_joint_idx = 0;
+	Transform3D target_global_pose = Transform3D();
+	Transform3D origin_global_pose = Transform3D();
+
+	void chain_backwards();
+	void chain_forwards();
+	void chain_apply();
+
+protected:
+	static void _bind_methods();
+	bool _get(const StringName &p_path, Variant &r_ret) const;
+	bool _set(const StringName &p_path, const Variant &p_value);
+	void _get_property_list(List<PropertyInfo> *p_list) const;
```

Comment:
Please consider to add `virtual` and `override`, this will allow the compiler to better guide us in case of refactoring:
```c++
	virtual bool _get(const StringName &p_path, Variant &r_ret) const override;
	virtual bool _set(const StringName &p_path, const Variant &p_value) override;
	virtual void _get_property_list(List<PropertyInfo> *p_list) const override;
```